tests/stack-snapshot-deps/hs_override_stack_test.go

```diff
@@ -0,0 +1,107 @@
+package hs_override_stack_test
+
+import (
+    it "github.com/tweag/rules_haskell/tests/integration_testing"
+    "github.com/bazelbuild/rules_go/go/tools/bazel_testing"
+    "strings"
+    "testing"
+)
+
+func TestMain(m *testing.M) {
+    it.TestMain(m, `
+-- WORKSPACE --
+local_repository(
+    name = "rules_haskell",
+    path = "../rules_haskell",
+)
+
+load("@rules_haskell//haskell:repositories.bzl", "rules_haskell_dependencies")
+
+rules_haskell_dependencies()
+
+load("@rules_haskell//haskell:nixpkgs.bzl", "haskell_register_ghc_nixpkgs")
+
+haskell_register_ghc_nixpkgs(
+    attribute_path = "haskell.compiler.ghc8107",
+    repository = "@rules_haskell//nixpkgs:default.nix",
+    version = "8.10.7",
+)
+
+load("@rules_haskell//haskell:toolchain.bzl", "rules_haskell_toolchains")
+
+rules_haskell_toolchains(version = "8.10.7")
+
+load(
+    "@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl",
+    "nixpkgs_cc_configure",
+    "nixpkgs_python_configure",
+)
+
+nixpkgs_cc_configure(
+    name = "nixpkgs_config_cc",
+    repository = "@rules_haskell//nixpkgs:default.nix",
+)
+
+nixpkgs_python_configure(
+    repository = "@rules_haskell//nixpkgs:default.nix",
+)
+
+load(
+    "@rules_haskell//haskell:cabal.bzl",
+    "stack_snapshot",
+    "use_stack",
+)
+
+# use dummy stack that only passes version check.
+# this should override the default behavior of 'stack_snapshot'
+# to use whatever is available in the environment.
+use_stack("//:stack")
+
+stack_snapshot(
+    name = "stackage",
+    local_snapshot = "@rules_haskell//:stackage_snapshot.yaml",
+    packages = [
+        "base",
+    ],
+)
+-- stack --
+#!/bin/sh
+echo 2.3.1
+
+-- BUILD.bazel --
+load(
+    "@rules_haskell//haskell:defs.bzl",
+    "haskell_test",
+)
+
+haskell_test(
+    name = "hs-bin",
+    srcs = ["Quux.hs"],
+    visibility = ["//visibility:public"],
+    deps = [ "@stackage//:base" ],
+)
+
+-- Quux.hs --
+module Main (main) where
+
+main :: IO ()
+main = putStrLn "Hello GHCi!"
+`)
+}
+
+func TestHsBinRepl(t *testing.T) {
+    // FIXME: This test is only partial due to limitations of `bazel_testing`.
+    // It is not possible to define a file as executable in `txtar`, and no
+    // obvious way to inject a pre-made executable into the build.
+    // To do this properly, the whole testing method needs to be reworked.
```

Comment:
I'm happy to leave this out of this PR and address in a follow-up to reduce scope, but, a good way to achieve this may be to define the binary in the top-level repository, i.e. somewhere under `@rules_haskell//tests/...`, and then make it a data dependency of the integration test. To that end you'll need to expose the [data attribute](https://github.com/tweag/rules_haskell/blob/102cf5ff1e2e0982c34c888cfd0555b2d9dd252b/tests/integration_tests.bzl#L21). Then it would be a pre-built file from the perspective of the integration test and could be used in `use_stack`.